src/can.rs

```diff
@@ -0,0 +1,357 @@
+/// but `filter` only accepts `0` in this bit position and `mask` says that this bit must be compared.
+///
+/// Accept only CAN ID `0x567`
+/// ```
+/// let filter = 0x567;
+/// // every bit must match filter
+/// let mask   = 0x7FF;
+/// let f = Filter::Standard { filter, mask };
+/// ```
+///
+/// Accept CAN IDs `0x560 - 0x56F`
+/// ```
+/// let filter = 0x560;
+/// // do not care about 4 LSB bits
+/// let mask   = 0x7F0;
+/// let f = Filter::Standard { filter, mask };
+/// ```
+pub enum Filter {
+    // Filter for 11 bit standard CAN IDs
+    Standard { filter: u16, mask: u16 },
+    // Filter for 29 bit extended CAN IDs
+    Extended { filter: u32, mask: u32 },
+}
+
+impl Filter {
+    /// Filter that allows all standard CAN IDs.
+    pub fn standard_allow_all() -> Self {
+        Self::Standard {
+            filter: 0,
+            mask: 0x7FF,
+        }
+    }
+
+    /// Filter that accepts all extended CAN IDs.
+    pub fn extended_allow_all() -> Self {
+        Self::Extended {
+            filter: 0,
+            mask: 0x1FFFFFFF,
+        }
+    }
+}
+
+/// CAN abstraction
+pub struct CanBus<TX: OutputPin, RX: InputPin> {
+    tx: PhantomData<TX>,
+    rx: PhantomData<RX>,
+}
+
+impl<TX: OutputPin, RX: InputPin> CanBus<TX, RX> {
+    pub fn new(tx: TX, rx: RX, timing: Timing, filter: Filter) -> Result<Self, EspError> {
+        let general_config = twai_general_config_t {
+            mode: twai_mode_t_TWAI_MODE_NORMAL,
+            tx_io: tx.pin(),
+            rx_io: rx.pin(),
+            clkout_io: -1,
+            bus_off_io: -1,
+            tx_queue_len: 5,
+            rx_queue_len: 5,
+            alerts_enabled: TWAI_ALERT_NONE,
+            clkout_divider: 0,
+            intr_flags: ESP_INTR_FLAG_LEVEL1 as i32,
+        };
+
+        let timing_config = timing.timing();
+
+        // modify filter and mask to be compatible with TWAI acceptance filter
+        let (filter, mask) = match filter {
+            Filter::Standard { filter, mask } => ((filter as u32) << 21, !((mask as u32) << 21)),
+            Filter::Extended { filter, mask } => (filter << 3, !(mask << 3)),
+        };
+
+        let filter_config = twai_filter_config_t {
+            acceptance_code: filter,
+            acceptance_mask: mask,
+            single_filter: true,
+        };
+
+        esp!(unsafe { twai_driver_install(&general_config, &timing_config, &filter_config) })?;
+        esp!(unsafe { twai_start() })?;
+
+        Ok(Self {
+            tx: PhantomData,
+            rx: PhantomData,
+        })
+    }
+}
+
+impl<TX: OutputPin, RX: InputPin> embedded_hal::can::blocking::Can for CanBus<TX, RX> {
+    type Frame = Frame;
+    type Error = embedded_hal::can::ErrorKind;
+
+    fn transmit(&mut self, frame: &Self::Frame) -> Result<(), Self::Error> {
+        esp!(unsafe { twai_transmit(&frame.0, portMAX_DELAY) })
+            .map_err(|_| embedded_hal::can::ErrorKind::Other)
```

Comment:
You are swallowing the original EspError here. Not really sure that's OK - I haven't looked (yet) into the new error types of embedded-hal V1.0. Just wanted to raise the flag. Will follow up on this and on the error handling (with `ErrorKind::Other`) later.